### PR TITLE
[PT 4] Local storage and settings

### PR DESCRIPTION
**Pull request by KauhareniaNastassia**
Add saving menu items to the local storage if internet connection is not available
Add saving shopping cart to the local storage
Add changing theme mode to settings page
Add changing color theme to settings page
Add links for contacts
Add permissions for links

https://github.com/KauhareniaNastassia/food_delivery_app/assets/104850152/e9ba97d3-7c93-47a5-8bac-7a5cfb62e4f9
https://github.com/KauhareniaNastassia/food_delivery_app/assets/104850152/a5171819-c3ed-4eee-be0b-fed6275ab369

### CODE REVIEW

File: core/lib/core.dart [hunk 1]
```diff
@@ -1,6 +1,12 @@
...
 export 'package:cloud_firestore/cloud_firestore.dart';
+export 'package:connectivity_plus/connectivity_plus.dart';
+export 'package:flutter_bloc/flutter_bloc.dart';
+export 'package:hive/hive.dart';
+export 'package:hive_flutter/hive_flutter.dart';
+export 'package:motion_toast/motion_toast.dart';
+export 'package:liquid_pull_to_refresh/liquid_pull_to_refresh.dart';
 export 'config/di/app_di.dart';
 export 'config/di/data_di.dart';
+export 'package:permission_handler/permission_handler.dart';
```
Comment (SashaPilevich) [inline core/lib/core.dart]: space

File: core_ui/lib/src/widgets/app_bar.dart [hunk 2]
```diff
@@ -1,39 +1,19 @@
...
   @override
   Widget build(BuildContext context) {
-    return BlocBuilder<AppThemeBloc, AppThemeState>(
-      builder: (BuildContext context, AppThemeState state) {
+    return BlocBuilder<SettingsBloc, SettingsState>(
+      builder: (BuildContext context, SettingsState state) {
         return AppBar(
           title: const Text('Food Delivery App'),
           backgroundColor: Theme.of(context).appBarTheme.backgroundColor,
           titleSpacing: Theme.of(context).appBarTheme.titleSpacing,
```
Comment (SashaPilevich) [inline core_ui/lib/src/widgets/app_bar.dart]: create variable

File: core_ui/lib/src/widgets/buttons/primary_button.dart [hunk 3]
```diff
@@ -13,21 +15,30 @@ class PrimaryButton extends StatelessWidget {
...
+          borderRadius: BorderRadius.all(
+            Radius.circular(15),
+          ),
         ),
       ),
       child: Container(
-        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 10),
+        padding: const EdgeInsets.symmetric(
+          horizontal: 10,
+          vertical: 10,
```
Comment (SashaPilevich) [inline core_ui/lib/src/widgets/buttons/primary_button.dart:34]: recomendation: create class for dimensions

File: data/lib/providers/settings_local_provider.dart [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    final Box colorScheme = await Hive.openBox('colorScheme');
+    colorScheme.put(
+      'colorScheme',
+      isStandard.toString(),
+    );
+  }
+
+  Future<double> getFontSizeFromLocal() async {
+    final Box colorScheme = await Hive.openBox('fontSize');
+    final fontSize = colorScheme.get('fontSize').toString();
```
Comment (SashaPilevich) [inline data/lib/providers/settings_local_provider.dart]: type

File: data/lib/repositories/menu_repository_impl/menu_repository_impl.dart [hunk 5]
```diff
@@ -1,24 +1,43 @@
...
+        _localMenuProvider = localMenuProvider;
 
   @override
   Future<List<MenuItemModel>> fetchMenuItems() async {
-    final List<MenuItemEntity> menuItems =
-        await _menuDataProvider.fetchMenuItems();
-    return menuItems
-        .map((MenuItemEntity e) => MenuItemMapper.toModel(e))
-        .toList();
+    final ConnectivityResult connectivityResult =
```
Comment (SashaPilevich) [inline data/lib/repositories/menu_repository_impl/menu_repository_impl.dart]: create class in core

File: features/main_page_view/lib/src/bloc/main_page/bloc.dart [hunk 6]
```diff
@@ -1,37 +1,72 @@
...
+import 'package:core/core.dart';
+import 'package:domain/domain.dart';
 
+part 'event.dart';
 part 'state.dart';
 
 class MenuBloc extends Bloc<MenuEvent, MenuState> {
   final FetchMenuItemsUseCase _fetchMenuItemsUseCase;
 
   MenuBloc({required FetchMenuItemsUseCase fetchMenuItemsUseCase})
```
Comment (SashaPilevich) [inline features/main_page_view/lib/src/bloc/main_page/bloc.dart]: ,

File: features/main_page_view/lib/src/bloc/main_page/state.dart [hunk 7]
```diff
@@ -1,19 +1,30 @@
...
+  MenuState copyWith({
+    List<MenuItemModel>? menu,
+    String? exception,
+    bool? isInternetConnectionAvailableState,
+    bool? isLoading,
+  }) {
+    return MenuState(
+      menu: menu ?? this.menu,
+      exception: exception ?? this.exception,
+      isInternetConnectionAvailableState: isInternetConnectionAvailableState ??
```
Comment (SashaPilevich) [inline features/main_page_view/lib/src/bloc/main_page/state.dart]: hasInternetConnection

File: features/main_page_view/lib/src/ui/components/main_page_screen.dart [hunk 8]
```diff
@@ -18,42 +19,71 @@ class _MainPageScreenState extends State<MainPageScreen> {
   @override
   Widget build(BuildContext context) {
     final Size size = MediaQuery.of(context).size;
```
Comment (SashaPilevich) [inline features/main_page_view/lib/src/ui/components/main_page_screen.dart]: final MediaQueryData mediaQueryData = MediaQuery.of(context);

File: features/main_page_view/lib/src/ui/components/main_page_screen.dart [hunk 9]
```diff
@@ -18,42 +19,71 @@ class _MainPageScreenState extends State<MainPageScreen> {
...
+              ).show(context);
+            }
+          },
+          builder: (BuildContext context, MenuState state) {
+            if (state.isLoading) {
+              return const LoadingIndicator();
+            }
+            if (state.menu.isNotEmpty) {
+              return PageRefresher(
+                onRefresh: () {
```
Comment (SashaPilevich) [inline features/main_page_view/lib/src/ui/components/main_page_screen.dart]: async

File: features/main_page_view/lib/src/ui/components/main_page_screen.dart [hunk 10]
```diff
@@ -18,42 +19,71 @@ class _MainPageScreenState extends State<MainPageScreen> {
...
+                  return Future<void>(
+                    () => context.read<MenuBloc>().add(
+                          InitEvent(),
+                        ),
```
Comment (SashaPilevich) [inline features/main_page_view/lib/src/ui/components/main_page_screen.dart]: create variable

File: features/main_page_view/lib/src/ui/components/menu_item_details_screen.dart [hunk 11]
```diff
@@ -84,7 +84,7 @@ class _MenuItemDetailsScreenState extends State<MenuItemDetailsScreen> {
                 child: Container(
                   padding: const EdgeInsets.fromLTRB(30, 0, 30, 40),
                   color: Theme.of(context).cardColor,
```
Comment (SashaPilevich) [inline features/main_page_view/lib/src/ui/components/menu_item_details_screen.dart]: create variable

File: features/main_page_view/lib/src/ui/components/widgets/menu_item/item_list_button.dart [hunk 12]
```diff
@@ -19,7 +21,7 @@ class ItemListButtonState extends State<ItemListButton> {
   @override
   Widget build(BuildContext context) {
     final Size size = MediaQuery.of(context).size;
```
Comment (SashaPilevich) [inline features/main_page_view/lib/src/ui/components/widgets/menu_item/item_list_button.dart]: same

File: features/main_page_view/lib/src/ui/components/widgets/menu_item/menu_item_title.dart [hunk 13]
```diff
@@ -14,6 +16,7 @@ class MenuItemTitle extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
     final Size size = MediaQuery.sizeOf(context);
```
Comment (SashaPilevich) [inline features/main_page_view/lib/src/ui/components/widgets/menu_item/menu_item_title.dart]: same

File: features/settings/lib/src/bloc/settings_bloc/bloc.dart [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    on<InitAppSettings>(_initAppSettings);
+    on<AppThemeChangingEvent>(_appThemeChanged);
+    on<AppColorSchemeChangingEvent>(_appColorSchemeChanged);
+    on<AppFontSizeChangingEvent>(_appFontSizeChanged);
+
+    add(InitAppSettings());
+  }
+
+  Future<void> _initAppSettings(
+      InitAppSettings event, Emitter<SettingsState> emit) async {
```
Comment (SashaPilevich) [inline features/settings/lib/src/bloc/settings_bloc/bloc.dart]: ,

File: features/settings/lib/src/ui/settings.dart [hunk 15]
```diff
@@ -1,30 +1,52 @@
...
+        return Scaffold(
+          body: Padding(
+            padding: const EdgeInsets.all(30),
+            child: Column(
+              mainAxisAlignment: MainAxisAlignment.start,
+              children: <Widget>[
+                SwitchToTheme(
+                  isLight: state.isLight,
+                  onTap: () {
+                    context.read<SettingsBloc>().add(
```
Comment (SashaPilevich) [inline features/settings/lib/src/ui/settings.dart]: create bloc variable

File: features/shopping_cart/lib/src/bloc/bloc.dart [hunk 16]
```diff
@@ -1,45 +1,78 @@
...
-        if (shoppingCartItems[i].menuItem == event.menuItem) {
-          shoppingCartItems[i].amount++;
-          itemIsInCart = true;
+      if (shoppingCartItems.isEmpty) {
+        emit(
+          state.copyWith(shoppingCart: state.shoppingCart),
+        );
+      } else {
+        double totalPrice = 0.0;
+        for (final item in shoppingCartItems) {
```
Comment (SashaPilevich) [inline features/shopping_cart/lib/src/bloc/bloc.dart]: type

File: core/lib/config/di/data_di.dart [hunk 17]
```diff
@@ -52,4 +85,74 @@ class DataDI {
...
+    );
+
+    instance.registerLazySingleton<RemoveShoppingCartItemUseCase>(
+      () => RemoveShoppingCartItemUseCase(
+        shoppingCartRepository: instance.get<ShoppingCartRepository>(),
+      ),
+    );
+  }
+
+  _initSettings() {
```
Comment (AliaksandrHuzich) [inline core/lib/config/di/data_di.dart]: void

File: core/lib/config/internet_connection_check/connection_di.dart [hunk 18]
```diff
@@ -0,0 +1,12 @@
+import 'package:core/core.dart';
```
Comment (AliaksandrHuzich) [inline core/lib/config/internet_connection_check/connection_di.dart]: file naming

Comment (AliaksandrHuzich) [reply core/lib/config/internet_connection_check/connection_di.dart]: to data as a provider/service

File: core_ui/lib/src/theme/app_theme.dart [hunk 19]
```diff
@@ -2,91 +2,148 @@ import 'package:core_ui/core_ui.dart';
...
-    /// navigation bar styles
-    bottomNavigationBarTheme: const BottomNavigationBarThemeData(
-      backgroundColor: AppColors.backgroundBottomNavigationBarColor,
-      selectedItemColor: AppColors.primaryColor,
-      unselectedItemColor: AppColors.unselectedItemColor,
-    ),
-  );
+  ThemeData get lightThemeData {
+    return ThemeData(
+      scaffoldBackgroundColor: AppColors.backgroundColor,
```
Comment (AliaksandrHuzich) [inline core_ui/lib/src/theme/app_theme.dart:26]: убрать условия сделать 4 темы или тернарным

File: core_ui/lib/src/widgets/page_refresher.dart [hunk 20]
```diff
@@ -0,0 +1,30 @@
+import 'package:core/core.dart';
+import 'package:flutter/material.dart';
+
+class PageRefresher extends StatelessWidget {
+  final Widget child;
+  final Future<void> Function() onRefresh;
```
Comment (AliaksandrHuzich) [inline core_ui/lib/src/widgets/page_refresher.dart:6]: VoidCallback()

File: data/lib/entity/shopping_cart_item/shopping_cart_item_entity.dart [hunk 21]
```diff
@@ -0,0 +1,11 @@
+import 'package:data/data.dart';
+
+class ShoppingCartItemEntity {
+  final MenuItemEntity menuItemEntity;
+  int amount;
```
Comment (AliaksandrHuzich) [inline data/lib/entity/shopping_cart_item/shopping_cart_item_entity.dart]: final

Comment (AliaksandrHuzich) [reply data/lib/entity/shopping_cart_item/shopping_cart_item_entity.dart]: for changing - copyWith

File: data/lib/providers/local_menu_provider.dart [hunk 22]
```diff
@@ -0,0 +1,24 @@
+import 'package:data/data.dart';
+import 'package:domain/domain.dart';
+
+class LocalMenuProvider {
+  LocalMenuProvider();
```
Comment (AliaksandrHuzich) [inline data/lib/providers/local_menu_provider.dart]: const

File: data/lib/providers/local_menu_provider.dart [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+
+  Future<void> saveMenuItemsToLocal(List<MenuItemModel> menuItems) async {
+    final Box<MenuItemEntity> menuItemsBox = await Hive.openBox('menuItems');
+    if (menuItemsBox.isEmpty) {
```
Comment (AliaksandrHuzich) [inline data/lib/providers/local_menu_provider.dart]: useless

File: data/lib/providers/local_shopping_cart_provider.dart [hunk 24]
```diff
@@ -0,0 +1,62 @@
+import 'package:data/data.dart';
+
+class LocalShoppingCartProvider {
+  LocalShoppingCartProvider();
```
Comment (AliaksandrHuzich) [inline data/lib/providers/local_shopping_cart_provider.dart]: const

File: data/lib/providers/local_shopping_cart_provider.dart [hunk 25]
```diff
@@ -0,0 +1,62 @@
+import 'package:data/data.dart';
```
Comment (AliaksandrHuzich) [inline data/lib/providers/local_shopping_cart_provider.dart]: rewrite this file (just a hive provider not a cart one)

File: domain/lib/usecases/add_shopping_cart_item_use_case.dart [hunk 26]
```diff
@@ -0,0 +1,13 @@
+import 'package:domain/domain.dart';
+
+class AddShoppingCartItemUseCase {
+  final ShoppingCartRepository _shoppingCartRepository;
+
+  const AddShoppingCartItemUseCase(
+      {required ShoppingCartRepository shoppingCartRepository})
```
Comment (AliaksandrHuzich) [inline domain/lib/usecases/add_shopping_cart_item_use_case.dart]: ,

File: features/main_page_view/lib/src/bloc/main_page/bloc.dart [hunk 27]
```diff
@@ -1,37 +1,70 @@
...
+      },
+    );
   }
 
   Future<void> _onLoadMenu(
     InitEvent event,
     Emitter<MenuState> emit,
   ) async {
-    emit(MenuLoadingState());
+    add(IsInternetConnectionAvailableEvent());
```
Comment (AliaksandrHuzich) [inline features/main_page_view/lib/src/bloc/main_page/bloc.dart]: remove

File: features/main_page_view/lib/src/bloc/main_page/bloc.dart [hunk 28]
```diff
@@ -1,37 +1,70 @@
...
+        state.copyWith(
+          exception: e.toString(),
+          isLoading: false,
+        ),
       );
     }
   }
+
+  Future<void> _isInternetConnectionAvailable(
+    IsInternetConnectionAvailableEvent event,
```
Comment (AliaksandrHuzich) [inline features/main_page_view/lib/src/bloc/main_page/bloc.dart:58]: передай резалт сюда

File: features/main_page_view/lib/src/ui/components/main_page_screen.dart [hunk 29]
```diff
@@ -17,43 +18,70 @@ class _MainPageScreenState extends State<MainPageScreen> {
...
-                children: <Widget>[
-                  const Column(
-                    crossAxisAlignment: CrossAxisAlignment.center,
-                    children: <Widget>[
-                      SizedBox(height: 20),
-                    ],
+        child: BlocConsumer<MenuBloc, MenuState>(
+          listener: (BuildContext context, MenuState state) {
+            if (!state.isInternetConnectionAvailable) {
+              MotionToast.error(
```
Comment (AliaksandrHuzich) [inline features/main_page_view/lib/src/ui/components/main_page_screen.dart]: to separate widget

File: features/main_page_view/lib/src/ui/components/main_page_screen.dart [hunk 30]
```diff
@@ -17,43 +18,70 @@ class _MainPageScreenState extends State<MainPageScreen> {
...
+                  child: Center(
+                    child: SingleChildScrollView(
+                      controller: _scrollController,
+                      child: Column(
+                        crossAxisAlignment: CrossAxisAlignment.center,
+                        children: <Widget>[
+                          const Column(
+                            crossAxisAlignment: CrossAxisAlignment.center,
+                            children: <Widget>[
+                              SizedBox(height: 20),
```
Comment (AliaksandrHuzich) [inline features/main_page_view/lib/src/ui/components/main_page_screen.dart:62]: todo

File: features/settings/lib/src/bloc/open_links_bloc/bloc.dart [hunk 31]
```diff
@@ -0,0 +1,115 @@
+import 'package:core/core.dart';
+import 'package:url_launcher/url_launcher.dart';
+
+part 'event.dart';
+part 'state.dart';
+
+class OpenContactLinksBloc
+    extends Bloc<OpenContactLinksEvent, OpenContactLinksState> {
+  OpenContactLinksBloc() : super(OpenContactLinksState()) {
```
Comment (AliaksandrHuzich) [inline features/settings/lib/src/bloc/open_links_bloc/bloc.dart]: formatting

File: features/settings/lib/src/bloc/open_links_bloc/bloc.dart [hunk 32]
```diff
@@ -0,0 +1,115 @@
...
+    } catch (e) {
+      emit(
+        ErrorOpenLinkState(
+          errorMessage: e.toString(),
+        ),
+      );
+    }
+  }
+
+  void _openMessengerLink(
```
Comment (AliaksandrHuzich) [inline features/settings/lib/src/bloc/open_links_bloc/bloc.dart]: Future

File: features/settings/lib/src/bloc/open_links_bloc/bloc.dart [hunk 33]
```diff
@@ -0,0 +1,115 @@
...
+    OpenMessengerEvent event,
+    Emitter<OpenContactLinksState> emit,
+  ) async {
+    const String telephoneNumber = '+375291234567';
+    final Uri messageUrl = Uri.parse('sms:$telephoneNumber');
+
+    PermissionStatus status = await Permission.sms.request();
```
Comment (AliaksandrHuzich) [inline features/settings/lib/src/bloc/open_links_bloc/bloc.dart]: final

File: features/settings/lib/src/ui/components/change_color_scheme.dart [hunk 34]
```diff
@@ -0,0 +1,85 @@
...
+
+    return Row(
+      mainAxisAlignment: MainAxisAlignment.spaceBetween,
+      children: <Widget>[
+        Text(
+          'Change color scheme',
+          style: theme.textTheme.titleLarge,
+        ),
+        GestureDetector(
+          onTap: !isStandardColorScheme ? onTap : () {},
```
Comment (AliaksandrHuzich) [inline features/settings/lib/src/ui/components/change_color_scheme.dart]: развернуть условие

File: features/settings/lib/src/ui/components/change_color_scheme.dart [hunk 35]
```diff
@@ -0,0 +1,85 @@
...
+                  width: 2,
+                ),
+              ),
+              child: !isStandardColorScheme
+                  ? Icon(
+                      Icons.check,
+                      size: size.width * 0.08,
+                      color: Colors.white,
+                    )
+                  : null),
```
Comment (AliaksandrHuzich) [inline features/settings/lib/src/ui/components/change_color_scheme.dart]: ,

**Pull request merged**